Equi template fails on macos and ios

**Describe the bug**
Equi example fails on all devices I have access to. Macos and ios.

Macos fails with

```rust
error[E0603]: function `main` is private
   --> gen/bin/desktop.rs:3:18
    |
3   |     mobile_equi::main();
    |                  ^^^^ private function
    |
note: the function `main` is defined here
   --> /Users/awk21/Projects/Coding/Rust/mobile_equi/src/lib.rs:160:1
    |
160 | fn main() {
    | ^^^^^^^^^

For more information about this error, try `rustc --explain E0603`.
error: could not compile `mobile-equi` due to previous error
```

and ios with

```
Undefined symbols for architecture arm64:
  "_start_app", referenced from:
      _main in main.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
note: Building targets in dependency order
note: Run script build phase 'Build Rust Code' will be run during every build because the option to run the script phase "Based on dependency analysis" is unchecked. (in target 'mobile-equi_iOS' from project 'mobile-equi')
** BUILD FAILED **

error: Failed to build via `xcodebuild`
    command ["xcodebuild"] exited with code 65
```

The first error is easy to fix with the addition of a pub but the latter needed more work


**Steps To Reproduce**
Create a equi tempalte and try to run or archive it with apple

**Expected behavior**
No errors

**Platform and Versions (please complete the following information):**
Host OS: Macos
Target OS: Macos and ios
Rustc: rustc 1.69.0 (84c898d65 2023-04-16)
`cargo mobile doctor`:
  
  [✔] cargo-mobile v0.5.0
      • Contains commits up to "Add egui example (#164)\n"
      • Installed at "~/.cargo/.tauri-mobile"
      • macOS v12.5 (21G72)
      • rustc v1.69.0 (84c898d65 2023-4-16)
  
  [✔] Apple developer tools
      • Xcode v14.2
      • Active developer dir: "/Applications/Xcode.app/Contents/Developer"
      • ios-deploy v1.12.2
      • XcodeGen v2.35.0
      ✗ xcode-rust-plugin plugin absent
      ✗ xcode-rust-plugin lang spec absent
      ✗ xcode-rust-plugin lang metadata absent
      • xcode-rust-plugin is up-to-date
      ✗ xcode-rust-plugin doesn't support Xcode UUID "C91F3560-00E7-4749-8E3F-4D83B1496051"
      • Development team: Andrew King (N7VJA8E4L8)
  
  [!] Android developer tools
      ✗ Have you installed the Android SDK? The `ANDROID_HOME` environment variable isn't set, and is required: environment variable not found
  
  [!] Connected devices
      ✗ Failed to get iOS device list: Failed to request device list from `ios-deploy`: command ["ios-deploy", "--detect", "--timeout", "1", "--json", "--no-wifi"] exited with code
        253

**Would you want to assign yourself to resolve this bug?**
- [x] Yes
- [ ] No

**Additional context**
See comments on #164 too

--- a/templates/apps/egui/src/lib.rs
+++ b/templates/apps/egui/src/lib.rs
@@ -139,6 +139,40 @@ fn _main(event_loop: EventLoop<Event>) {
     });
 }
 
+#[cfg(any(target_os = "ios", target_os = "android"))]
+fn stop_unwind<F: FnOnce() -> T, T>(f: F) -> T {
+    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(f)) {
+        Ok(t) => t,
+        Err(err) => {
+            eprintln!("attempt to unwind out of `rust` with err: {:?}", err);
+            std::process::abort()
+        }
+    }
+}
+
+#[cfg(target_os = "ios")]
+fn _start_app() {
+    stop_unwind(|| main());
+}
+
+#[no_mangle]
+#[inline(never)]
+#[cfg(target_os = "ios")]
+pub extern "C" fn start_app() {
+    _start_app();
+}
+
+#[cfg(not(target_os = "android"))]
+pub fn main() {
+    env_logger::builder()
+        .filter_level(log::LevelFilter::Warn)
+        .parse_default_env()
+        .init();
+
+    let event_loop = EventLoopBuilder::with_user_event().build();
+    _main(event_loop);
+}
+
 #[allow(dead_code)]
 #[cfg(target_os = "android")]
 #[no_mangle]
@@ -152,17 +186,5 @@ fn android_main(app: AndroidApp) {
     let event_loop = EventLoopBuilder::with_user_event()
         .with_android_app(app)
         .build();
-    _main(event_loop);
-}
-
-#[allow(dead_code)]
-#[cfg(not(target_os = "android"))]
-fn main() {
-    env_logger::builder()
-        .filter_level(log::LevelFilter::Warn)
-        .parse_default_env()
-        .init();
-
-    let event_loop = EventLoopBuilder::with_user_event().build();
-    _main(event_loop);
+    stop_unwind(|| _main(event_loop));
 }

